### PR TITLE
Implement topk_transform_ragged with SYCL

### PR DESCRIPTION
**Pull request by cyxlily**
## Motivation

Add SYCL implementations of `topk_transform_ragged` on XPU.
This op is required by
`sglang.kernels.ops.attention.dsv4.topk.topk_transform_ragged_v2`(https://github.com/sgl-project/sglang/blob/main/python/sglang/kernels/ops/attention/dsv4/topk.py#L90), which is the
prefill-path entry point used by the DSA top-k backend.
Based on the CUDA implementation: https://github.com/sgl-project/sglang/blob/main/python/sglang/kernels/jit/csrc/deepseek_v4/topk_v2.cuh

## Modifications

- **New op: `topk_transform_ragged`**
  - Added `TopKTransformRaggedKernel` (SYCL) and its host launcher in
    `src/sycl/TopKTransform.cpp`, reusing the existing `radix_topk` /
    `trivial_topk_transform_ragged` helpers with a *runtime* `topk` (inferred
    from `out_indices.size(1)`), unlike the topk=2048-fixed
    `fast_topk_transform_ragged_fused`.
  - Selects the top-k within each row's `[row_start, row_start + seq_len)`
    window and rebases the selected positions onto the flattened KV via
    `out_offsets`, `-1`-padding the rest — matching the semantics of the CUDA
    `TopKKernel::transform_ragged` reference.
  - Declared in `include/sgl_kernel_ops.h` and registered as
    `topk_transform_ragged(Tensor scores, Tensor seq_lens, Tensor! out_indices,
    Tensor out_offsets, Tensor? row_starts) -> ()` in
    `src/torch_extension_sycl.cc`.
  - Added the Python wrapper in `python/sgl_kernel/top_k.py` and exported it
    from `python/sgl_kernel/__init__.py`.

- **Tests** (`tests/test_topk_transform.py`)
  - Added `test_topk_transform_ragged`, `test_topk_transform_ragged_length_trivial`,
    and `test_topk_transform_ragged_length_ragged`, covering fixed-length,
    trivial (`length <= topk`), and mixed trivial/radix-path rows, with and
    without `row_starts`.
  - Added `test_topk_transform_paged_page_table_none` to cover the
    `page_tables=None` (raw-index output) mode of `topk_transform_paged`,
    which previously had no test coverage.

- **Benchmark** (`benchmark/bench_topk_transform.py`)
  - Added `benchmark_topk_transform_ragged` and its config sweep, wired into
    the script's `__main__` run/report block.

- **Naming cleanup**
  - Renamed `topk_transform_512` → `topk_transform` and `topk_transform_512_v2` →
    `topk_transform_paged` to align with the CUDA op names.

## Accuracy Tests

Tested topk_transform_ragged (SYCL) vs PyTorch reference on Intel XPU.
All unit tests passed.

Command: pytest tests/test_topk_transform.py -k topk_transform_ragged
Results: 
========================================================================== test session starts ==========================================================================
platform linux -- Python 3.12.3, pytest-9.1.1, pluggy-1.6.0
rootdir: /home/lily/PR/topk_transform_ragged/sgl-kernel-xpu
configfile: pyproject.toml
plugins: anyio-4.14.2
collected 766 items / 602 deselected / 164 selected

tests/test_topk_transform.py .................................................................................................................................... [ 80%]
................................                                                                                                                                  [100%]

============================================================ 164 passed, 602 deselected in 66.36s (0:01:06) =============================================================

## Performance

Hardware: Intel BMG. Median of 20 iterations, 5 warmup, torch.xpu.Event
timing.
SYCL is faster than the Torch reference across all reported configurations.

topk_transform_ragged: sycl vs torch
|   bs |   topk |   seq_len | row_starts |   sycl_ms |   torch_ms |   speedup |
|-----:|-------:|----------:|-----------:|----------:|-----------:|----------:|
|  132 |    512 |      4096 |      False |  0.155469 |   0.454115 |     2.921 |
|  132 |   1024 |      4096 |      False |  0.152187 |   0.450000 |     2.957 |
|  132 |   2048 |      4096 |      False |  0.111667 |   0.448385 |     4.015 |
|  132 |    512 |     16384 |       True |  0.384323 |   2.169218 |     5.644 |
|  132 |   1024 |     16384 |       True |  0.391875 |   2.197500 |     5.608 |
|  132 |   2048 |     16384 |       True |  0.407292 |   2.197708 |     5.396 |
|  132 |    512 |     65536 |      False |  1.360416 |   7.459115 |     5.483 |
|  132 |   1024 |     65536 |      False |  1.364323 |   7.489948 |     5.490 |
|  132 |   2048 |     65536 |      False |  1.364687 |   7.514740 |     5.507 |
|  256 |    512 |     16384 |      False |  0.664792 |   3.768073 |     5.668 |
|  256 |   1024 |     16384 |      False |  0.673281 |   3.803750 |     5.650 |
|  256 |   2048 |     16384 |      False |  0.704531 |   3.880000 |     5.507 |
| 4096 |    512 |      4096 |      False |  3.703594 |   7.167604 |     1.935 |
| 4096 |   1024 |      4096 |      False |  3.627291 |   8.113802 |     2.237 |
| 4096 |   2048 |      4096 |      False |  2.575313 |   9.850260 |     3.825 |
| 4096 |    512 |     16384 |       True |  9.760521 |  64.557708 |     6.614 |
| 4096 |   1024 |     16384 |       True |  9.888177 |  65.601979 |     6.634 |
| 4096 |   2048 |     16384 |       True | 10.285886 |  67.226042 |     6.536 |

### CODE REVIEW

File: benchmark/bench_topk_transform.py [hunk 1]
```diff
@@ -351,10 +360,9 @@ def benchmark_topk_transform_512_v2(bs, topk, seq_len, page_size, provider):
         .contiguous()
     )
     out_page = torch.full((bs, topk), -1, dtype=torch.int32, device="xpu")
-    metadata = torch.empty((bs + 1, 2), dtype=torch.int32, device="xpu")
 
-    fn = lambda: topk_transform_512_v2(
-        score, lengths, page_table, out_page, page_size, metadata, None
+    fn = lambda: topk_transform_paged(
+        score, lengths, page_table, out_page, page_size, None
     )
```
[the remark this answers is not in the document]
Comment (cyxlily) [reply benchmark/bench_topk_transform.py:367]: Thanks for the comments, updated.

File: src/sycl/TopKTransform.cpp [hunk 2]
```diff
@@ -786,24 +854,77 @@ void topk_transform_512_launch(
...
+  TORCH_CHECK(scores.stride(1) == 1, "scores must have unit last-dim stride");
+  TORCH_CHECK(scores.scalar_type() == at::kFloat, "scores must be float32");
+
+  const int64_t B = scores.size(0);
+
+  TORCH_CHECK(seq_lens.dim() == 1 && seq_lens.is_contiguous() && seq_lens.size(0) == B);
+  TORCH_CHECK(seq_lens.scalar_type() == at::kInt, "seq_lens must be int32");
+
+  TORCH_CHECK(out_offsets.dim() == 1 && out_offsets.is_contiguous() && out_offsets.size(0) == B);
+  TORCH_CHECK(out_offsets.scalar_type() == at::kInt, "out_offsets must be int32");
```
[the remark this answers is not in the document]
Comment (cyxlily) [reply src/sycl/TopKTransform.cpp:902]: It's consistent with the CUDA reference (topk_v2.cuh), which relies on the same "caller-guaranteed" contract without an explicit bounds check.

File: src/sycl/TopKTransform.cpp [hunk 3]
```diff
@@ -786,24 +854,77 @@ void topk_transform_512_launch(
...
+  topk_transform_paged_launch(scores, seq_lens, page_tables_opt, out_page_indices, page_size, std::nullopt);
+}
+
+SGL_KERNEL_EXPORT void topk_transform_ragged(
+    const at::Tensor& scores,
+    const at::Tensor& seq_lens,
+    at::Tensor& out_indices,
+    const at::Tensor& out_offsets,
+    std::optional<at::Tensor> row_starts_opt) {
+  CHECK_INPUT(scores);
```
Comment (mingfeima) [inline src/sycl/TopKTransform.cpp]: `CHECK_INPUT(scores)` requires a contiguous tensor.

The sibling paged launch only checks device + `stride(1)==1` and indexes with `stride(0)`.

CUDA v2 ragged likewise accepts a padded view (`stride(0) > width`).

Prefill indexer logits in sglang are often that layout

`CHECK_DEVICE` + last-dim contiguous, not `is_contiguous()`

**Review by mingfeima — APPROVED** [review]
some minor changes needed. generally LGTM.

File: src/sycl/TopKTransform.cpp (hunk 3 above)
Comment (cyxlily) [reply src/sycl/TopKTransform.cpp]: Thanks for the comment, updated.
